### PR TITLE
Add aws_cloudwatch_alarm resource

### PR DESCRIPTION
**Pull request by clintoncwolfe**
This adds very basic support for a singular Cloudwatch Alarm resource, based on searching for alarms by metric name / namespace.  

Only one matcher implemented: `exists?`

Only one property implemented: `alarm_actions`.

Docs, unit tests, and integration tests included.

### CODE REVIEW

File: test/unit/resources/aws_cloudwatch_alarm_test.rb [hunk 1]
```diff
@@ -0,0 +1,167 @@
+require 'ostruct'
+require 'helper'
+require 'aws_cloudwatch_alarm'
+
+# MCWAB = MockCloudwatchAlarmBackend
+# Abbreviation not used outside this file
+
+#=============================================================================#
+#                            Constructor Tests
```
Comment (jerryaldrichiii) [inline test/unit/resources/aws_cloudwatch_alarm_test.rb]: In InSpec core we've historically we've used the `.must` ( e.g. `foo.must_equal` ) format instead of `assert`. For the sake of consistency, we might need to rewrite these. Thoughts?

File: libraries/aws_cloudwatch_alarm.rb [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+    search
+  end
+
+  def exists?
+    @exists
+  end
+
+  private
+
+  def validate_resource_params(raw_params)
```
[the remark this answers is not in the document]
Comment (clintoncwolfe) [reply libraries/aws_cloudwatch_alarm.rb]: I agree; I'm just waiting on #118 to be accepted.  Until it is, I'm submitting PRs in a pre-#118 world, with tighter restrictions.

File: libraries/aws_cloudwatch_alarm.rb [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+    unless raw_params.is_a? Hash
+      raise ArgumentError, "Resource params should be passed using \"key: 'value'\" format."
+    end
+
+    # Currently you must specify exactly metric_name and metric_namespace
+    [:metric_name, :metric_namespace].each do |param|
+      raise ArgumentError, "Missing resource param #{param}" unless raw_params.key?(param)
+      instance_variable_set(:"@#{param}", raw_params.delete(param))
```
[the remark this answers is not in the document]
Comment (clintoncwolfe) [reply libraries/aws_cloudwatch_alarm.rb]: I agree, after I wrote it it "smelled bad".  It's also inconsistent with other resources.  I'll return the validated args, then set in the constructor.

File: test/unit/resources/aws_cloudwatch_alarm_test.rb (hunk 1 above)
Comment (clintoncwolfe) [reply test/unit/resources/aws_cloudwatch_alarm_test.rb]: So, the inspec-aws project has been using dependency injection along with assertions, see #110 .  Work is currently being done to remove the DI stuff, which leaves it using the Minitest-class+assertions approach.  

My anticipation is that in a later issue, we'll cut over to using expectations-style (`must`), as used by inspec-core.  We could decide that all new resources (including this one) use expectations.  

I'm happy to make the change, just need a sense of when to make that decision.

File: test/integration/verify/controls/aws_cloudwatch_alarm.rb [hunk 4]
```diff
@@ -0,0 +1,19 @@
+alarm_01 = attribute(
+  'cloudwatch_alarm_01',
+  default: 'default.cloudwatch_alarm',
+  description: 'Name of Cloudwatch Alarm')
+
+  # TODO: this is likely to be a multiple hit
```
[the remark this answers is not in the document]
Comment (clintoncwolfe) [reply test/integration/verify/controls/aws_cloudwatch_alarm.rb]: I suppose the right thing to do is watch a metric that we ourselves created ... which we create on the LMF PR.  So I think I should reword it accordingly.

File: libraries/aws_cloudwatch_alarm.rb (hunk 3 above)
[the remark this answers is not in the document]
Comment (chris-rock) [reply libraries/aws_cloudwatch_alarm.rb]: I agree, a method that is called `validator` should not change anything. I like the specific check for all required attributes, though. Setting the instance variables in the initialize is a good way forward.

File: libraries/aws_cloudwatch_alarm.rb [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+      raise 'More than one Cloudwatch Alarm was matched. Try using ' \
+        "more specific resource parameters. Alarms matched: #{alarms.join(', ')}"
+    else
+      @alarm_actions = aws_alarms.metric_alarms.first.alarm_actions
+      @alarm_name = aws_alarms.metric_alarms.first.alarm_name
+      @exists = true
+    end
+  end
+
+  class Backend
```
Comment (chris-rock) [inline libraries/aws_cloudwatch_alarm.rb]: I am not 100% I am understanding the reason for that abstraction. Can you share your reasoning?

File: test/integration/verify/controls/aws_cloudwatch_alarm.rb (hunk 4 above)
[the remark this answers is not in the document]
Comment (chris-rock) [reply test/integration/verify/controls/aws_cloudwatch_alarm.rb]: I have never seen NOTE, is that a new one?

File: test/unit/resources/aws_cloudwatch_alarm_test.rb (hunk 1 above)
Comment (chris-rock) [reply test/unit/resources/aws_cloudwatch_alarm_test.rb]: Is there any reason why we should not use `assert`? We are not proposing the syntax for InSpec users, because it is only understood well by programmers and the should syntax is more readable. I do not see an issue why we should not use this as part of our unit tests. I also think that we do not need to standardize on that, since both ways are useful depending on what you are trying to implement.

**Review by chris-rock — COMMENTED** [review]
Great work @clintoncwolfe Left some comments, but nothing major.

File: test/integration/verify/controls/aws_cloudwatch_alarm.rb (hunk 4 above)
[the remark this answers is not in the document]
Comment (jerryaldrichiii) [reply test/integration/verify/controls/aws_cloudwatch_alarm.rb]: I've never seen `NOTE` either. Might just reword it to be an actionable statement.

File: libraries/aws_cloudwatch_alarm.rb (hunk 5 above)
Comment (clintoncwolfe) [reply libraries/aws_cloudwatch_alarm.rb]: In the unit tests, I need to be able to have something that acts like AWS, but actually isn't.  A typical unit test will subclass SomeResource::Backend, then implement the interface, typically using lookup tables to provide dummy data for the tests.  The unit test then calls SomeResource::Backend.select(MyDummyBackend), and we've effectively swapped the provider.

I use the same pattern in #112 and #116.  Other solutions are possible; if we don't like this one, we should get together and design another.

Some resources were mocking the backend using classes passed in through the constructor (#110) which seemed more problematic, since the constructor is part of the public interface of the resource used by control authors.

File: test/unit/resources/aws_cloudwatch_alarm_test.rb (hunk 1 above)
Comment (adamleff) [reply test/unit/resources/aws_cloudwatch_alarm_test.rb]: @chris-rock I see an issue only in terms of consistency. I'd rather all our unit tests look-and-feel the same, and knowing this is going to be merged into core InSpec, I think we need to think carefully about maintaining a consistent style.

**Review by chris-rock — APPROVED** [review]
Thank you @clintoncwolfe

**Comment by chris-rock** [comment]
@clintoncwolfe can you rebase and resolve the conflicts?
